### DISCUSSION TITLE
Question about Pinentry

### DISCUSSION THREAD

**Discussion by rashil2000** [git-for-windows/git#3616] [Q&A]
Hi all!

Has anyone from the community succeeded in using pinentry-tty or pinentry-ncurses for GPG commit signing? On my Linux boxes, whenever I make a commit a terminal-based prompt for password comes up. But on Windows, there is a GUI version that one types the password into.

According to my own (limited) testing, I see that Git only ships with a GUI version of pinentry.exe (pinentry-w32.exe, I'm not sure if this is GTK or Qt built). As a result, I have built pinentry-tty.exe and pinentry-ncurses.exe from source (instructions here - https://github.com/gpg/pinentry), but I can't figure out how to configure the gpg-agent to use the ttry (or ncurses) version. Searching on StackOverflow only gives Linux-specific answers (some of which I tried, but to no avail).

One use case of this is signing commits over an SSH session (Windows host).

**Comment by rimrul** [comment]
> pinentry-w32.exe, I'm not sure if this is GTK or Qt built

Neither. [Windows API](https://github.com/gpg/pinentry/tree/master/w32).

> As a result, I have built pinentry-tty.exe and pinentry-ncurses.exe from source (instructions here - https://github.com/gpg/pinentry), but I can't figure out how to configure the gpg-agent to use the ttry (or ncurses) version. Searching on StackOverflow only gives Linux-specific answers (some of which I tried, but to no avail).

Looking at the `gpg-agent` source, I think [this one should work](https://superuser.com/a/521027), but you might need to build an `MSys2` version of `pinentry`.

**Reply by rashil2000** [reply]
> Neither. [Windows API](https://github.com/gpg/pinentry/tree/master/w32).

Ah okay, that makes more sense.

> Looking at the `gpg-agent` source, I think [this one should work](https://superuser.com/a/521027)

I tried this. I exported GPG_TTY by doing `export GPG_TTY=$(tty)` and then:

With pinentry-tty, I get this:

![image](https://user-images.githubusercontent.com/46838874/148593013-e6b896c4-377e-4d90-af9b-1ab593670f71.png)

With pinentry-curses, I get this:

![image](https://user-images.githubusercontent.com/46838874/148593486-a156a609-c784-44a1-a48e-9d540ceab947.png)

> but you might need to build an `MSys2` version of `pinentry`.

I built it some time ago, but I think it's the MSYS2 version: 

![image](https://user-images.githubusercontent.com/46838874/148592436-53e7974d-0891-4f7e-b086-f89345bc6b9b.png)

**Reply by rimrul** [reply]
That does indeed look like MSYS2 versions. 
The fact you're getting two different error messages probably means it's at least picking up the config and trying to do something based on it.
But I'm not sure how we could convince gpg to give us some more meaningful error messages.

**Reply by rashil2000** [reply]
It is possible to print a log file (ref: https://github.com/diablodale/pinentry-wsl-ps1/issues/6). This is my ~/.gnupg/gpg-agent.conf:

```
pinentry-program /usr/local/bin/pinentry-curses
debug 1024
debug-pinentry
log-file ~/.cache/agent-curses.log
```

I can share the log files. But they seem to contain keys/hashes and stuff. Would you think it's safe to paste the file here?

**Reply by rashil2000** [reply]
Upon closer inspection, I find that the log file does not contain any useful information (that's different from pinentry-w32), apart from the error messages already shown in the screenshots.
